File: iosApp/iosApp/Features/HomeGreetingCarousel.swift
// HomeGreetingCarousel.swift
// Rotating welcome carousel on the Home tab — replaces the static
// "Hi {firstName} 👋" line. Sourced from CustomerDashboardViewModel's
// `headlinePrefix` + `greetings` StateFlows (built in the KMP shared
// module so iOS and Android render identical copy).
//
// Behaviour:
//   - Auto-advance every 5 seconds with a 0.35s ease-in-out fade-cross.
//   - Pauses while the user is long-pressing the card.
//   - Single greeting → renders static (no rotation).
//   - Tap → pushes the destination view onto the Home stack and calls
//     `markGreetingSeen(id:)` so the freshness rule dismisses the
//     greeting on next emission.

import SwiftUI
import ThapsusShared

struct HomeGreetingCarousel: View {
    let vm: CustomerDashboardViewModel?
    /// Called when the carousel taps a greeting whose destination is
    /// `HomeGreetingDestination.NpsSurvey`. The host view owns the sheet
    /// state and presents `NpsSurveyView` — surveys aren't a push.
    var onNpsTap: () -> Void = {}

    @State private var prefixObs: StateFlowObserver<String>? = nil
    @State private var greetingsObs: StateFlowObserver<[HomeGreeting]>? = nil

    @State private var index: Int = 0
    @State private var paused: Bool = false

    private let rotateInterval: Duration = .seconds(5)
    private let fade: Animation = .easeInOut(duration: 0.35)

    var body: some View {
        VStack(alignment: .leading, spacing: 6) {
            Text(prefixObs?.value ?? "hi.")
                .font(.body(15, weight: .semibold))
                .foregroundStyle(LG.fg3)
                .accessibilityAddTraits(.isHeader)

            content
        }
        .padding(.top, 8)
        .padding(.bottom, 4)
        .frame(maxWidth: .infinity, alignment: .leading)
        .onAppear { bootstrap() }
        .task { await rotationLoop() }
    }

    // MARK: - Content

    @ViewBuilder
    private var content: some View {
        let list = currentGreetings
        if let current = list[safe: index] {
            interactiveGreeting(current)
                .simultaneousGesture(
                    LongPressGesture(minimumDuration: 0.15)
                        .onChanged { _ in paused = true }
                        .onEnded { _ in paused = false }
                )

            if list.count > 1 {
                HStack(spacing: 6) {
                    ForEach(0..<list.count, id: \.self) { i in
                        Capsule()
                            .fill(i == index ? LG.fg.opacity(0.85) : LG.fg.opacity(0.20))
                            .frame(width: i == index ? 18 : 6, height: 4)
                            .animation(fade, value: index)
                    }
                }
                .padding(.top, 2)
            }
        } else {
            // Pre-bootstrap fallback so the layout doesn't pop in.
            Text("ready when you are.")
                .font(.display(24, weight: .heavy))
                .foregroundStyle(LG.fg)
        }
    }

    @ViewBuilder
    private func interactiveGreeting(_ current: HomeGreeting) -> some View {
        // NPS surveys are a sheet, not a stack push. Branch here so the
        // host view can mount `NpsSurveyView` via its own state binding.
        if current.destination is HomeGreetingDestinationNpsSurvey {
            Button {
                vm?.markGreetingSeen(greetingId: current.id)
                onNpsTap()
            } label: {
                greetingText(current)
            }
            .buttonStyle(.plain)
        } else {
            NavigationLink {
                current.destination.makeDestinationView()
                    .onAppear { vm?.markGreetingSeen(greetingId: current.id) }
            } label: {
                greetingText(current)
            }
            .buttonStyle(.plain)
        }
    }

    private func greetingText(_ current: HomeGreeting) -> some View {
        Text(current.body)
            .font(.display(24, weight: .heavy))
            .foregroundStyle(LG.fg)
            .multilineTextAlignment(.leading)
            .fixedSize(horizontal: false, vertical: true)
            .frame(maxWidth: .infinity, alignment: .leading)
            .id(current.id)
            .transition(.opacity.combined(with: .move(edge: .bottom)))
    }

    // MARK: - State helpers

    private var currentGreetings: [HomeGreeting] {
        greetingsObs?.value ?? []
    }

    private func bootstrap() {
        guard let vm else { return }
        if prefixObs == nil {
            prefixObs = StateFlowObserver(initial: vm.headlinePrefix.value) {
                vm.headlinePrefix
            }
        }
        if greetingsObs == nil {
            greetingsObs = StateFlowObserver(initial: vm.greetings.value) {
                vm.greetings
            }
        }
    }

    private func rotationLoop() async {
        while !Task.isCancelled {
            try? await Task.sleep(for: rotateInterval)
            if paused { continue }
            let count = currentGreetings.count
            guard count > 1 else { continue }
            withAnimation(fade) {
                index = (index + 1) % count
            }
        }
    }
}

// MARK: - Safe array subscript

private extension Array {
    subscript(safe i: Int) -> Element? {
        return (i >= 0 && i < count) ? self[i] : nil
    }
}

File: androidApp/src/main/kotlin/com/thapsus/cargo/android/ui/customer/HomeGreetingCarousel.kt
package com.thapsus.cargo.android.ui.customer

import androidx.compose.animation.AnimatedContent
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.togetherWith
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.thapsus.cargo.android.ui.theme.Brand
import com.thapsus.cargo.presentation.CustomerDashboardViewModel
import com.thapsus.cargo.presentation.home.HomeGreeting
import com.thapsus.cargo.presentation.home.HomeGreetingDestination
import kotlinx.coroutines.delay

/**
 * Rotating welcome carousel on the Customer Home tab — replaces the static
 * "Welcome, {firstName}" header. Sourced from `headlinePrefix` + `greetings`
 * StateFlows on [CustomerDashboardViewModel] (built in the KMP shared module
 * so iOS and Android render identical copy).
 *
 * Behaviour:
 * - Auto-advances every 5 seconds with a fade cross-transition.
 * - Resets to index 0 whenever the source list changes — keeps the most
 *   urgent greeting visible after a refresh.
 * - Single greeting renders static (no rotation, no page indicator).
 * - Tap fires [onTap] with the current greeting's destination and calls
 *   [CustomerDashboardViewModel.markGreetingSeen] so status variants drop
 *   on the next emission.
 */
@Composable
fun HomeGreetingCarousel(
    dashVm: CustomerDashboardViewModel,
    onTap: (HomeGreetingDestination) -> Unit,
    modifier: Modifier = Modifier
) {
    val prefix by dashVm.headlinePrefix.collectAsStateWithLifecycle()
    val greetings by dashVm.greetings.collectAsStateWithLifecycle()

    var index by remember { mutableStateOf(0) }

    // Reset to top of the freshly emitted list — important when the urgent
    // ladder shifts (e.g. user just paid an invoice; the carousel should
    // start from whatever's now most urgent rather than mid-rotation).
    LaunchedEffect(greetings) {
        if (index >= greetings.size) index = 0
    }

    // 5-second rotation. Only runs when there's more than one greeting,
    // and uses `greetings.size` in the loop guard so adding/removing
    // entries reactively pauses or starts the loop.
    LaunchedEffect(greetings.size) {
        if (greetings.size <= 1) return@LaunchedEffect
        while (true) {
            delay(5_000)
            if (greetings.isEmpty()) break
            index = (index + 1) % greetings.size
        }
    }

    val current = greetings.getOrNull(index)

    Column(
        modifier = modifier
            .fillMaxWidth()
            .padding(top = 4.dp),
        verticalArrangement = Arrangement.spacedBy(6.dp)
    ) {
        Text(
            text = prefix.ifBlank { "hi." },
            color = Brand.ink.copy(alpha = 0.6f),
            fontWeight = FontWeight.SemiBold,
            fontSize = 14.sp
        )

        AnimatedContent(
            targetState = current,
            label = "home-greeting-carousel",
            transitionSpec = { fadeIn() togetherWith fadeOut() }
        ) { greeting ->
            if (greeting != null) {
                Text(
                    text = greeting.body,
                    color = Brand.ink,
                    fontWeight = FontWeight.ExtraBold,
                    fontSize = 26.sp,
                    lineHeight = 32.sp,
                    modifier = Modifier
                        .fillMaxWidth()
                        .clickable {
                            dashVm.markGreetingSeen(greeting.id)
                            onTap(greeting.destination)
                        }
                )
            } else {
                // Pre-bootstrap fallback — the layout shouldn't pop in.
                Text(
                    text = "ready when you are.",
                    color = Brand.ink,
                    fontWeight = FontWeight.ExtraBold,
                    fontSize = 26.sp,
                    lineHeight = 32.sp
                )
            }
        }

        if (greetings.size > 1) {
            Row(
                horizontalArrangement = Arrangement.spacedBy(6.dp),
                modifier = Modifier.padding(top = 2.dp)
            ) {
                repeat(greetings.size) { i ->
                    val active = i == index
                    Box(
                        modifier = Modifier
                            .height(4.dp)
                            .width(if (active) 18.dp else 6.dp)
                            .background(
                                color = if (active) Brand.ink.copy(alpha = 0.85f)
                                else Brand.ink.copy(alpha = 0.20f),
                                shape = RoundedCornerShape(2.dp)
                            )
                    )
                }
            }
        }
        Spacer(Modifier.height(2.dp))
    }
}

/**
 * Translates a [HomeGreetingDestination] from the KMP shared layer to the
 * appropriate [CustomerRoutes] string. The carousel hands the destination
 * to the scaffold which dispatches via this helper.
 *
 * Per-record deep-links (specific invoice / BFM order / ticket) fall back
 * to the corresponding list screen for v1 — the customer reaches the
 * record with one more tap from there. A follow-up can route through to
 * detail screens once those are wired with their own deep-link routes.
 */
fun HomeGreetingDestination.toCustomerRoute(): String = when (this) {
    is HomeGreetingDestination.Shop -> CustomerRoutes.SHOP
    is HomeGreetingDestination.ActivityHub -> CustomerRoutes.ACTIVITY
    is HomeGreetingDestination.Parcels -> CustomerRoutes.TRACKING
    is HomeGreetingDestination.Invoices -> CustomerRoutes.INVOICES
    is HomeGreetingDestination.Transactions -> CustomerRoutes.TRANSACTIONS
    is HomeGreetingDestination.Consolidations -> CustomerRoutes.CONSOLIDATIONS
    is HomeGreetingDestination.CreditCenter -> CustomerRoutes.CREDIT
    is HomeGreetingDestination.BuyForMeOrder -> CustomerRoutes.SHOP
    is HomeGreetingDestination.PayInvoice -> CustomerRoutes.INVOICES
    is HomeGreetingDestination.TicketDetail -> CustomerRoutes.TICKETS
    is HomeGreetingDestination.Dsar -> CustomerRoutes.DSAR
    is HomeGreetingDestination.Referral -> CustomerRoutes.REFERRAL
    is HomeGreetingDestination.NpsSurvey -> CustomerRoutes.ACTIVITY
}
